### PR TITLE
Stochastic Inference for Dropout

### PR DESCRIPTION
**Pull request by xternalz**
Applying dropout stochastically during inference, as an alternative to the 'expectation' approach.

### CODE REVIEW

File: Dropout.lua [hunk 1]
```diff
@@ -20,13 +21,14 @@ function Dropout:updateOutput(input)
       self.output:resizeAs(input):copy(input)
    end
    if self.p > 0 then
-      if self.train then
-         self.noise:resizeAs(input)
-         self.noise:bernoulli(1-self.p)
+      local si = self.si and not self.train
+      if self.train or si and not self.v2 then
+         local noise = si and input:clone() or self.noise:resizeAs(input)
```
Comment (fmassa) [inline Dropout.lua]: Do you need to clone the input at every iteration? Also, the previous `self.noise:resizeAs(input)` is not adapted in this situation?

Comment (xternalz) [inline Dropout.lua]: This helps to save memory during testing when self.noise is cleared/uninitialized.

File: Dropout.lua [hunk 2]
```diff
@@ -20,13 +21,14 @@ function Dropout:updateOutput(input)
       self.output:resizeAs(input):copy(input)
    end
    if self.p > 0 then
-      if self.train then
-         self.noise:resizeAs(input)
-         self.noise:bernoulli(1-self.p)
+      local si = self.si and not self.train
+      if self.train or si and not self.v2 then
```
Comment (fmassa) [inline Dropout.lua]: Are you sure about this condition?
Just a few lines downstream in the same `if` block, there is the case where `v2` is used considered, but it will never reach it with the current condition.

Comment (xternalz) [inline Dropout.lua]: It will reach the block for `v2` when `self.train == true`.

File: Dropout.lua (hunk 1 above)
Comment (fmassa) [inline Dropout.lua]: If you want to save memory during inference, it is usually better to just share the `noise` field among other dropout layers or other internal buffers from other modules.
Currently, this line is actually allocating new memory at every iteration, and freeing it just afterwards, which is a blocking operation on GPU. We usually try to avoid that as much as possible in `nn`.

Comment (xternalz) [inline Dropout.lua]: Thanks for pointing this out. I will remove the cloning and use just `self.noise`. Has the feature of sharing `noise` field among layers been implemented in your [optnet](https://github.com/fmassa/optimize-net) library?

File: Dropout.lua [hunk 3]
```diff
@@ -1,10 +1,11 @@
 local Dropout, Parent = torch.class('nn.Dropout', 'nn.Module')
 
-function Dropout:__init(p,v1,inplace)
+function Dropout:__init(p,v1,inplace,stochasticInference)
    Parent.__init(self)
    self.p = p or 0.5
    self.train = true
    self.inplace = inplace
+   self.si = stochasticInference or false
```
Comment (apaszke) [inline Dropout.lua]: `si` is a very cryptic name. It would be better to keep it more verbose.

**Comment by soumith** [comment]
please document this option

**Comment by soumith** [comment]
hi @xternalz . This is good to go, except that it looks like it needs a rebase on top of master.

**Comment by joeyhng** [comment]
Would it be clearer if removing ```and not self.train``` which seems not needed?

Also, can this be done by simply setting the module in evaluate mode?

**Comment by xternalz** [comment]
Please check again @soumith.

File: Dropout.lua [hunk 4]
```diff
@@ -20,7 +21,7 @@ function Dropout:updateOutput(input)
       self.output:resizeAs(input):copy(input)
    end
    if self.p > 0 then
-      if self.train then
+      if self.train or self.stochastic_inference and not self.train and not self.v2 then
```
Comment (soumith) [inline Dropout.lua]: hi, just looked at the logic again. isn't this wrong?
if it is evaluate mode, this will never come here because self.v2 will almost always be true.

shouldn't it just be:

if self.train or self.stochastic_inference

File: SpatialDropout.lua [hunk 5]
```diff
@@ -1,15 +1,16 @@
...
    self.p = p or 0.5
    self.train = true
+   self.stochastic_inference = stochasticInference or false
    self.noise = torch.Tensor()
 end
 
 function SpatialDropout:updateOutput(input)
    self.output:resizeAs(input):copy(input)
-   if self.train then
+   if self.train or self.stochastic_inference and not self.train then
```
Comment (soumith) [inline SpatialDropout.lua]: this can also simply be:

if self.train or self.stochastic_inference

**Comment by soumith** [comment]
added final comments. sorry for not catching these earlier.

File: Dropout.lua (hunk 4 above)
Comment (xternalz) [inline Dropout.lua]: Yea, you are right.

**Comment by soumith** [comment]
Thanks @xternalz !

**Comment by tastyminerals** [comment]
Guys, shouldn't `Dropout.lua` in **rnn** package be the same `Dropout.lua` in **nn**? **Element-Research** have updated it to include `mono` option. So, `LSTM.lua`, `FastLSTM.lua`, `GRU.lua` all have `mono` option which they expect for `Dropout` to accept:

**nn** `Dropout:__init(p,v1,inplace,stochasticInference)`
**rnn** `Dropout:__init(p,v1,inplace,lazy,mono)`
